traitsui/tests/editors/test_list_editor.py

```diff
@@ -170,6 +177,36 @@ def test_index_out_of_range(self):
             with self.assertRaises(IndexError):
                 people_list.locate(Index(10))
 
+    # regression test for enthought/traitsui#1154
+    @requires_toolkit([ToolkitName.qt])
+    @unittest.skipIf(no_modal_dialog_tester, "ModalDialogTester unavailable")
+    def test_add_item_fails(self):
+        from traits.api import TraitError
+
+        class Foo(HasStrictTraits):
+            dirs = List(Directory(exists=True))
+
+        obj = Foo()
+        tester = UITester(auto_process_events=False)
+        with tester.create_ui(obj) as ui:
+            dirs_list_editor = tester.find_by_name(ui, "dirs")
+
+            def trigger_error():
+                try:
+                    dirs_list_editor._target.add_empty()
+                except TraitError:
+                    return False
+                return True
+
+            mdtester = ModalDialogTester(trigger_error)
+            mdtester.open_and_run(lambda x: x.close())
+            # we want an error dialog to open, but don't want to raise a
+            # TraitError and crash the full application
+            self.assertTrue(mdtester.dialog_was_opened)
+            self.assertTrue(mdtester.result)
+
+            self.assertEqual(len(obj.dirs), 0)
+
```

Comment:
This test is a little ugly but it ultimately does what I wanted it to.

The behavior this PR capitalizes on is that if a trait is set to an invalid value, https://github.com/enthought/traitsui/blob/1c880889f19abbda50deddd497f7a24721b3a13a/traitsui/qt4/editor.py#L72-L90 will be called.  This is good as it gives an error dialog to the user saying why they can't do what they just did.  However, we don't really want this to crash the application if it happens when trying to add a new item to the list, so we swallow the exception there, and the list will just be left unchanged to how it was before.